Let people choose which payment methods are available to customers on order cycles

cillian opened the pull request:
#### What? Why?

For https://github.com/openfoodfoundation/openfoodnetwork/issues/8971, it uses the exact same approach as used in https://github.com/openfoodfoundation/openfoodnetwork/issues/8974.

Notes:
* This removes the `OpenFoodNetwork::AvailablePaymentMethodFilter` class and incorporates it into `OrderAvailablePaymentMethods`by using a new `Spree::PaymentMethod#configured?` method. I noticed the Enterprise.ready_for_checkout scope, which calls `Spree::PaymentMethod.available` doesn't check the things that used to be checked in `OpenFoodNetwork::AvailablePaymentMethodFilter` i.e. `.available` will return Stripe payment methods that are not configured with a Stripe account id.
#### What should we test?

I **think** the same acceptance and criteria as https://github.com/openfoodfoundation/openfoodnetwork/issues/8974 except replacing shipping methods with payment methods.

#### Release notes

Changelog Category: User facing changes

#### Documentation updates

The [Order Cycles (for hubs)](https://guide.openfoodnetwork.org/basic-features/shopfront/order-cycle/order-cycles-for-hubs) page in the user guide will probably need to be updated.

mkllnk commented on app/services/order_cycle_clone.rb:
```diff
@@ -0,0 +1,46 @@
+    oc.name = I18n.t("models.order_cycle.cloned_order_cycle_name", order_cycle: oc.name)
+    oc.orders_open_at = oc.orders_close_at = oc.mails_sent = oc.processed_at = nil
+    oc.coordinator_fee_ids = @original_order_cycle.coordinator_fee_ids
+    # rubocop:disable Layout/LineLength
+    oc.preferred_product_selection_from_coordinator_inventory_only = @original_order_cycle.preferred_product_selection_from_coordinator_inventory_only
+    # rubocop:enable Layout/LineLength
```
```suggestion
    oc.preferred_product_selection_from_coordinator_inventory_only =
      @original_order_cycle.preferred_product_selection_from_coordinator_inventory_only
```

mkllnk reviewed (approved):
Great work, thank you!

I can only offer a couple of style suggestions. :wink: 

The two spec fails seem to be unrelated and flaky.

cillian commented on spec/models/spree/payment_method_spec.rb:
```diff
@@ -5,99 +5,165 @@
-            description: 'foofah',
-            distributors: [enterprise]
-          )
+        Spree::Config.set(stripe_connect_enabled: true)
```
Thanks, good idea, that should be fixed by https://github.com/openfoodfoundation/openfoodnetwork/pull/9755/commits/8fb88eefa0bdf80ecc0b02ac0e8eac2e349c00cf now. Let me know if you need me to squash anything.

mkllnk reviewed (approved):
Ah, great, thank you. After I approved the pull request I remembered that this caching thing could be a real pain when we try to parallelise the specs. So using these stubs will probably save us some work there.

jibees reviewed (approved):
It's though, and I've nothing to add. Thanks for this work @cillian ! 💪 🙏

drummer83 commented:
Hi @cillian,
Thanks for the next big piece of this work! :muscle: 
I have created a separate testing document with more details about the findings below (marked with :x:): https://docs.google.com/document/d/1eniZlEB4rxbU4PQSzeE2sKRPJcuQD18TtwdDIwb6ouc/edit?usp=sharing

I have tested this with split checkout activated because it's the future of our platform. I have found a few things that we should probably address before merging though.

**General findings regarding user experience**
- Leaving step 3 of complex order cycle asks to confirm to leave the page even when no data was changed.
- Leaving step 4 of complex order cycle DOESN’T ask to confirm to leave the page even when data was changed (the change is lost without notice).
- The buttons ‘save’ and ‘save and back to list’ should only be active when data was changed.
- If no payment methods (or shipping methods) are selected and the order cycle is saved, then ALL payment/shipping methods are activated automatically. Is this intended or should there be a warning instead?

**Acceptance Criteria**
1. All existing order cycles should have all payment methods available as now by default. :heavy_check_mark: 
2. When an order cycle is created all payment methods are selected and available by default. :heavy_check_mark: 
3. In the simple order cycle screen nothing changes :heavy_check_mark: 
4. ~~In the edit order cycle screen for hubs (complex) an additional expandable panel in step 3 of the Edit OC process specifies the number of selected and available payment methods when closed (5/6). Payment methods not available in the shopfront should be excluded from these numbers.~~ (outdated)
5. When expanded this panel should display any Payment Methods that are available. Payment methods that are available via the shopfront (e.g. not 'backoffice only' payment methods) appear with a checkbox enabling them to be activated/deactivated for the order cycle (ticked means activated). 'Backoffice only' payment methods should be 'greyed out' and unticked in this view. :heavy_check_mark: 
6. When only a subset of payment methods are selected, only selected payment methods should be visible to choose on this order cycle in the shopfront. Unselected payment methods will no longer be visible on checkout.
  - for a single distributor order cycle :heavy_check_mark: 
  - for a multi-distributor order cycle :x: 
7. All payment methods should remain available when creating subscriptions and back office orders. :heavy_check_mark: 
8. This feature has a higher importance than tagging when deciding which OCs to display e.g. if a cu10stomer is tagged for a payment method, the payment method must be available in the OC for it to be displayed. If this acceptance criteria is confusing let me know, because the requirement is actually simple but hard to explain in writing. :heavy_check_mark: 
9. If no payment methods are available the order cycle should not display. This also applies in the case of tagged shoppers, that might have a further reduced set of payment methods and hence visible order cycles. :x:
10. ~~The 'Ready For' field will move as per the image above to create more horizontal space for the additional expandable panel.~~ (outdated)
11. ~~The 'Customer Instructions' field will be moved to within the Shipping Methods expandable panel as per the image above.~~ (outdated)
12. Test the db migration :arrow_right: **OPEN**

My personal proposal would be to fix these issues before merging. But I may be too strict sometimes, so pinging @filipefurtad0, @RachL and @lin-d-hop to ask for their opinion. Adding feedback-needed for now.

Looking forward to finishing this up and thanks again!

filipefurtad0 commented:
Hey @drummer83 ,

Thanks for the thorough testing.

- On point 6., for multi-distributor OC: was this working for shipping methods, before this PR?
- On point 9., what happens if you complete checkout? Does it break checkout or rather attributes a random payment method? To put it in another way, what would be the severity here, if we choose to open an issue instead?

cillian commented:
Hi @drummer83 thanks for testing that and all the details.

I have fixed the issue in point (9) and added tests with this commit https://github.com/openfoodfoundation/openfoodnetwork/pull/9755/commits/eb72ec8e133288b5d826b39422566b8127da3c55, nice catch. 

However I can't seem to recreate the issue in point (6), to investigate further I might need access to staging.

drummer83 commented:
@filipefurtad0 

>    - On point 6., for multi-distributor OC: was this working for shipping methods, before this PR?

No, I could reproduce this in master as well.

>    -  On point 9., what happens if you complete checkout? Does it break checkout or rather attributes a random payment method? To put it in another way, what would be the severity here, if we choose to open an issue instead?

The customer gets stuck on the payment method step, because there is the error message to select one before moving on, but there are no options available - so you can't continue checkout.

![image](https://user-images.githubusercontent.com/1790176/197333666-3c64e13b-bdbc-4e2e-87f9-ce94b064b5de.png)

drummer83 commented:
Hi @cillian,

it took a a bit to figure out how to reproduce the issue (6) myself. It seems like the checkout isn't considering which shop I am actually shopping from. Instead the combination of available shipping methods is offered for selection. Some examples below, shopping at 'Konrad Non-Producer Hub', so only those options should be offered:

![image](https://user-images.githubusercontent.com/1790176/197333333-e2016fd9-663c-4be9-8b0e-299404548422.png)
![image](https://user-images.githubusercontent.com/1790176/197333337-68f1b6de-dd48-4a38-ad28-4b5ba876b3e5.png)
![image](https://user-images.githubusercontent.com/1790176/197333343-a3db3d0d-d744-4567-b085-d2f66047e5dd.png)

The same is true for **payment methods**.

Regarding **tagging**:
I tested the following settings for a customer allowed to see Paypal as payment option (on the LEFT) and a customer NOT allowed to see Paypal as payment option (on the RIGHT).
![image](https://user-images.githubusercontent.com/1790176/197334627-8a3be9ab-224e-45f6-94ed-bd4bc762b3e1.png)

Observations:

1. The shop appears to the customer NOT allowed to use Paypal as closed and open at the same time... Even though there is the 'shopfront closed' message, it is possible to put products in the cart and proceed to checkout. --> This normal behavior and OFN's way to create a display only shopfront (see [user guide](https://guide.openfoodnetwork.org/basic-features/shopfront/display-only-order-cycles)). I think we are good here and this wasn't being thought of when writing the specs. :heavy_check_mark: 
![image](https://user-images.githubusercontent.com/1790176/197335755-bb10b3da-6bf2-41aa-8ba7-14de84041ee3.png)

2. However the customer NOT allowed to use Paypal can't choose a payment method and gets stuck at this point during checkout. This is normal behavior in this case as well. :heavy_check_mark: 
![image](https://user-images.githubusercontent.com/1790176/197335132-eb7736ae-a130-4ad4-8dbf-4442a51df5f3.png)

## Summary
- We need to fix (6) for payment methods and shipping methods. :x:
- Tagging seems to work fine now! :heavy_check_mark: 
- Testing with legacy checkout is open - will do that once we are good on split checkout. :question: 
- Testing the migration is still open. :question: 

## Further possible improvements (outside of this PR)

- Leaving step 3 of complex order cycle asks to confirm to leave the page even when no data was changed.
- Leaving step 4 of complex order cycle DOESN’T ask to confirm to leave the page even when data was changed (the change is lost without notice).
- The buttons ‘save’ and ‘save and back to list’ should only be active when data was changed.
- If no payment methods (or shipping methods) are selected and the order cycle is saved, then ALL payment/shipping methods are activated automatically:
  - Is this intended or should there be a warning instead?

drummer83 commented:
Moved to In Dev to have another round of code review on this important functionality.

RachL commented:
Regarding point 6: if i understood correctly, in production, this won't work either for shipping methods? So perhaps we can merge this PR as is, and create another issue for this problem?

mkllnk reviewed (approved):
Looking good.

drummer83 commented:
Hi @cillian,
Alright, testing the remaining bits.

**Legacy checkout:**
- tagging works the same way as for split checkout (Delivery and Paypal have been tagged in the picture below)
- we also see issue (6), not further investigated here

![image](https://user-images.githubusercontent.com/1790176/197775824-2eb68dca-b070-41ee-9af8-c013f76481c0.png)

**Migration:**
- looking at the database we do see the new table

```
openfoodnetwork=> select * from order_cycles_distributor_payment_methods;
 order_cycle_id | distributor_payment_method_id 
----------------+-------------------------------
           3607 |                           866
           3607 |                           872
           3607 |                           873
           3607 |                           875

```

```
                                            Table "public.distributors_payment_methods"
      Column       |            Type             | Collation | Nullable |                         Default                          
-------------------+-----------------------------+-----------+----------+----------------------------------------------------------
 distributor_id    | integer                     |           |          | 
 payment_method_id | integer                     |           |          | 
 id                | bigint                      |           | not null | nextval('distributors_payment_methods_id_seq'::regclass)
 created_at        | timestamp without time zone |           |          | 
 updated_at        | timestamp without time zone |           |          | 


                                            Table "public.distributors_shipping_methods"
       Column       |            Type             | Collation | Nullable |                          Default                          
--------------------+-----------------------------+-----------+----------+-----------------------------------------------------------
 id                 | integer                     |           | not null | nextval('distributors_shipping_methods_id_seq'::regclass)
 distributor_id     | integer                     |           |          | 
 shipping_method_id | integer                     |           |          | 
 created_at         | timestamp without time zone |           | not null | 
 updated_at         | timestamp without time zone |           | not null | 

```

**Question:** Is there any reason why 'id' is a bigint for payment methods and an int for shipping methods? Could it be beneficial to have the same format at a later stage?

I think we are ready to merge this one but I will let @jibees and @mkllnk have a look at my test results and do the merge.

Apart from that I will create new issues for the remaining issue (6) and the things from 'further possible improvements' mentioned above.

mkllnk commented:
> Is there any reason why 'id' is a bigint for payment methods and an int for shipping methods? Could it be beneficial to have the same format at a later stage?

Good question. Rails changed the default from int to bigint at some point. And the payment method relationship was introduced much later than the shipping method relationship. Standardising these could be good at some point but it doesn't matter at the moment.

mkllnk commented:
Yes, good to go. Just awaiting confirmation from @RachL and @lin-d-hop that this can be merged with open issue 6:

* https://github.com/openfoodfoundation/openfoodnetwork/issues/9873

RachL commented:
This is not making any regression in production:I see no reason to hold on merging :)
